keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,338 @@
+}
+```
+
+We will discuss how the `StorageVersion` is updated and used in the [Changes to API
+servers](#changes-to-api-servers) section.
+
+## Changes to API servers
+
+### Curating a list of participating API servers in HA master
+
+Currently, API servers maintain such a list in the "kubernetes" endpoints. See
+this [design doc][] for details.
+
+We will formalize the API in another KEP. For the purpose of this KEP, all we
+need to know is that HA master is able to maintain a list of participating API
+servers in etcd.
+
+[design doc]:https://github.com/kubernetes/community/pull/939
+
+### Voting for Storage Version
+
+During bootstrap, for each resource, the kube-apiserver 
+* gets the StorageVersion object for this resource,
+  * if the object does not exist, creates one, setting this kube-apiserver's
+  preferred storage version as the SelectedVersion, also adding its preferred
+  storage version to the CandidateVersions list. Then jumps to the last step
+  (installing the resource handler).
+* gets the list of participating API servers,
+* updates the StorageVersion locally. Specifically,
+  * creates or updates the CandidateVersions, to express this kube-apiserver's
+    preferred storage version.
+  * checks if there is any version candidate whose APIServerID is not a
+    participating API server. This means the API server has gone and the entry
+    is stale. Removes such entries.
+  * checks if all participating API servers vote for the same version. If so,
+    sets the version as the SelectedVersion.
+* updates the StorageVersion object, using the resourceVersion in the first step
+  to avoid conflicting with other API servers.
+* installs the resource handler. Uses the SelectedVersion as the storage
+  version.
+
+We need to make sure that an API server establishes its membership in the list
+of participating API servers before reporting its preferred storage version via
+the StorageVersion API. Then the above procedure makes it impossible to mistake
+a new CandidateVersions entry as a stale one.
+
+### Voting in action
+
+**Scenario 1**, rolling upgrading HA API servers to a release that starts using
+the storage version voting.
+
+In this case, the first upgraded API server sets its preferred storage version as
+the selected version. Because the other not-upgraded-yet API servers do not
+respect the voting system, they might use a different version as the storage
+version. When they get upgraded, they will use the selected version as the
+storage version.
+
+**Scenario 2**, rolling upgrading HA API servers that have already enabled the
+storage version voting.
+
+In this case, the API servers will keep using the old selected storage version
+until all instances are upgraded to the new release. The last upgraded API
+server will change the selected version to the new storage version, and all API
+servers start to use the new version.
+
+### The selected storage version is supported by all API servers during upgrades/downgrades
```

Comment:
This describes selecting which version to use for a type which all API servers support, but at different versions.

Would the same mechanism correctly handle a type which some API servers do not support at all (e.g. avoid serving a new beta type while there are joined API servers that do not make that type available, xref https://github.com/kubernetes/kubernetes/issues/81286)

Reply:
I think so. It's mentioned in the "future work" section: "For example,
kube-apiservers can vote on what API is served, and only serve the API that is
supported by all servers."